### PR TITLE
Fix Issue 21288 & 10886 - incorrect this pointer

### PR DESCRIPTION
**Pull request by RazvanN7**
It seems that when a function is called the, if it requires a context pointer, the compiler simply rewrites it to `this.fcall` even though the function that is being call sometimes explicitly provides it's aggregate. This leads to bogus messages and also it makes it impossible to take the address of such functions if inside a this context.

I am fixing this by requiring that the this pointers of the called function and the outer aggregate match.

### CODE REVIEW

File: compiler/src/dmd/expressionsem.d [hunk 1]
```diff
@@ -1155,6 +1155,43 @@ L1:
...
+ *
+ * Returns:
+ *  `true` if outerFunc and calledFunc may use the same `this` pointer.
+ * `false` otherwise.
+ */
+private bool haveSameThis(FuncDeclaration outerFunc, FuncDeclaration calledFunc)
+{
+    auto thisAd = outerFunc.isMemberLocal();
+    if (!thisAd)
+        return false;
```
Comment (SixthDot) [inline compiler/src/dmd/expressionsem.d:1171]: It seems that this cant be covered because you call the new function pnly for members funcs, e.g line 7247.

```d
if (memberFunc && haveSameThis(memberFunc, exp.f))
```

so you could assert on `thisAd` instead

File: compiler/src/dmd/expressionsem.d [hunk 2]
```diff
@@ -1155,6 +1155,43 @@ L1:
...
+
+    auto requiredAd = calledFunc.isMemberLocal();
+    if (!requiredAd)
+        return false;
+
+    if (thisAd == requiredAd)
```
Comment (SixthDot) [inline compiler/src/dmd/expressionsem.d:1177]: wouldn't just `is` be enough here ?

File: compiler/src/dmd/expressionsem.d (hunk 1 above)
Comment (RazvanN7) [reply compiler/src/dmd/expressionsem.d:1171]: Yes, I was aware of that, but I figured that this function could be used from other contexts so it should not make assumptions about the status at the call site.

File: compiler/src/dmd/expressionsem.d (hunk 2 above)
Comment (RazvanN7) [reply compiler/src/dmd/expressionsem.d:1177]: Indeed, `is` would be preferred since we want to compare the pointers, but AST nodes do not define opEquals so we are fine with the default opEquals of object.

Long story-short: both is and == are fine in this case.

File: compiler/test/compilable/test19295.d [hunk 3]
```diff
@@ -0,0 +1,10 @@
+struct S1(T...) {
+    auto fun() {
+        static assert(__traits(compiles, &T[0]));
```
Comment (dkorpel) [inline compiler/test/compilable/test19295.d:3]: This compiles now, but does the resulting delegate have the right `this` parameter?

File: compiler/src/dmd/expressionsem.d [hunk 4]
```diff
@@ -1155,6 +1155,43 @@ L1:
...
+/*
+ * Check whether `outerFunc` and `calledFunc` have the same `this`.
+ * If `calledFunc` is the member of a base class of the class that contains
+ * `outerFunc` we consider that they have the same this.
+ *
+ * Returns:
+ *  `true` if outerFunc and calledFunc may use the same `this` pointer.
+ * `false` otherwise.
+ */
+private bool haveSameThis(FuncDeclaration outerFunc, FuncDeclaration calledFunc)
```
Comment (dkorpel) [inline compiler/src/dmd/expressionsem.d:1167]: This function doesn't look right. The logic in the function is surprisingly complex (many branches). I would expect `haveSameThis` to be commutative and take two arbitrary functions, but the names `outerFunc` and `calledFunc` suggest there are restrictions to the parameters which aren't documented.

File: compiler/src/dmd/expressionsem.d [hunk 5]
```diff
@@ -1155,6 +1155,43 @@ L1:
     return e1;
 }
 
+/*
+ * Check whether `outerFunc` and `calledFunc` have the same `this`.
+ * If `calledFunc` is the member of a base class of the class that contains
+ * `outerFunc` we consider that they have the same this.
```
Comment (dkorpel) [inline compiler/src/dmd/expressionsem.d:1161]: The example makes it unclear what "have same `this`" means, it sounds more like "have covariant `typeof(this)`"

File: compiler/src/dmd/expressionsem.d [hunk 6]
```diff
@@ -6892,8 +6930,6 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
         AggregateDeclaration ad = f.isMemberLocal();
         if (f.needThis())
             e.e1 = getRightThis(e.loc, sc, ad, e.e1, f);
-        if (e.e1.op == EXP.error)
-            return setError();
```
Comment (dkorpel) [inline compiler/src/dmd/expressionsem.d:6896]: Why is this removed?

**Review by dkorpel — APPROVED** [review]
My gut reaction to the diff was "this doesn't look right", but after reading it more carefully and reviewing it, it does look like a valid code change. Perhaps the documentation on `haveSameThis` could be clarified to make it easier to understand what's going on.

File: compiler/src/dmd/expressionsem.d (hunk 6 above)
Comment (RazvanN7) [reply compiler/src/dmd/expressionsem.d:6896]: Because it was added to fix a symptom of the bug this PR is fixing. Take a look at: https://github.com/dlang/dmd/pull/10123

File: compiler/test/compilable/test19295.d (hunk 3 above)
Comment (RazvanN7) [reply compiler/test/compilable/test19295.d:3]: But the result is not a delegate. It's a function, so there is no this pointer.

File: compiler/src/dmd/expressionsem.d (hunk 4 above)
Comment (RazvanN7) [reply compiler/src/dmd/expressionsem.d:1167]: Initially, I thought the fix is easy. I thought it would suffice to make sure that the this of the function containing the call and the this of the function being called are the same, however that does not cut it for derived member functions. Later on, I noticed that for nested functions there is some extra engineering done to get the right this. That's when I ended up with this function. I suspect that there are other places where this function might be used.

I don't think it can be made any simpler than that, but I am confident that what is in there at this point is correct.

File: compiler/src/dmd/expressionsem.d (hunk 5 above)
Comment (RazvanN7) [reply compiler/src/dmd/expressionsem.d:1161]: Well, I followed the existing naming pattern after `getRightThis`.

**Comment by RazvanN7** [comment]
@dkorpel I added some clarifications.

**Comment by RazvanN7** [comment]
@dkorpel Thanks for the review.

File: compiler/src/dmd/expressionsem.d (hunk 1 above)
Comment (SixthDot) [reply compiler/src/dmd/expressionsem.d:1195]: ok

File: compiler/src/dmd/expressionsem.d (hunk 2 above)
Comment (SixthDot) [reply compiler/src/dmd/expressionsem.d:1201]: I dont see the problem with `is`, for the comment about coverage you are protective, i.e "in case of" for but this one you are not.

**Comment by SixthDot** [comment]
This commit has introduced a regression, see https://issues.dlang.org/show_bug.cgi?id=24013.

**Comment by kinke** [comment]
I *guess* this also caused regression https://issues.dlang.org/show_bug.cgi?id=24109.

File: compiler/src/dmd/expressionsem.d [hunk 7]
```diff
@@ -1155,6 +1155,69 @@ L1:
...
+ *          X.gun();  // error
+ *      }
+ * }
+ *
+ * When `fun` is called, `outerfunc` = `sun` and `calledFunc = `fun`.
+ * `sun` is a member of `A` and `fun` is also a member of `A`, therefore
+ * `this` can be prepended to `fun`. When `gun` is called (it will result
+ * in an error, but that is not relevant here), which is a member of `X`,
+ * no `this` is needed because the outer function does not have the same
+ * `this` as `gun`.
```
Comment (WalterBright) [inline compiler/src/dmd/expressionsem.d:1185]: But it will fail anyway because X.gun() does not supply a `this` to `gun`.
